### ISSUE TITLE
Tensorflow Lite conversion output vs. Tensorflow output

### ISSUE THREAD

**Issue by gsirocco** [tensorflow/tensorflow#58206]
### 1. System information

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  Ubuntu 20.04.4 LTS (GNU/Linux 5.4.0-128-generic x86_64)
- TensorFlow installation (pip package or built from source): pip package
- TensorFlow library (version, if pip package or github SHA, if built from source): 2.9.1

### 2. Code

Provide code to help us reproduce your issues using one of the following options:

#### Option B: Paste your code here or provide a link to a custom end-to-end colab

```
import tensorflow as tf
from tensorflow import keras
import pickle
import numpy as np

with open("genMat.pkl", "rb") as f:
     gen_matrix = pickle.load(f)
# 14400x1800
xlen=14400
ylen=1800

def representative_dataset():
  for _ in range(100):
      data = np.random.randint(0, 2, (xlen,))
      yield [data.astype(np.float32)]

test_input = np.loadtxt("input_msg.csv",dtype=int,delimiter=',')
input = 0
generator = tf.convert_to_tensor(gen_matrix[input*xlen:(input+1)*xlen, :], dtype=tf.float32)

ldpc_enc_input = keras.Input(shape=(xlen,), name="ldpc_enc_input")
generator_input = tf.reshape(ldpc_enc_input, shape=(-1, xlen))
generator_output = tf.matmul(generator_input, generator)

ldpc_enc_output = tf.math.floormod(generator_output, 2)

ldpc_enc_model = keras.Model(inputs=ldpc_enc_input, outputs=ldpc_enc_output)

arr=ldpc_enc_model(test_input[input*xlen:(input+1)*xlen])
print(arr.numpy()[0])

converter = tf.lite.TFLiteConverter.from_keras_model(ldpc_enc_model)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.representative_dataset = representative_dataset
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.target_spec.supported_types = [tf.int8]
converter.inference_input_type = tf.int8 # or tf.uint8
converter.inference_output_type = tf.int8 # or tf.uint8

tflite_quant_model = converter.convert()
# Load the TFLite model and allocate tensors.
interpreter = tf.lite.Interpreter(model_content=tflite_quant_model)
interpreter.allocate_tensors()

# Get input and output tensors.
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

input_shape = input_details[0]['shape']
input_data = np.array(test_input[input*xlen:(input+1)*xlen], dtype=np.int8, ndmin=2)
interpreter.set_tensor(input_details[0]['index'], input_data)

interpreter.invoke()

# The function `get_tensor()` returns a copy of the tensor data.
# Use `tensor()` in order to get a pointer to the tensor.
arr = interpreter.get_tensor(output_details[0]['index'])
print(arr[0])
```  

[input_msg.csv](https://github.com/tensorflow/tensorflow/files/9834888/input_msg.csv)

### 3. Failure after conversion
[genMat.txt](https://github.com/tensorflow/tensorflow/files/9834891/genMat.txt)

If the conversion is successful, but the generated model is wrong, then state what is wrong:

- Model produces wrong results vs. the Tensorflow model. I have attached the pkl file as a text file which gets loaded and then the input csv gets loaded (input_msg.csv). The output of the TF invocation if correct and printed out ([0. 1. 0. ... 1. 0. 1.]). The TFLite invocation output is printed and is incorrect ([-128  127  127 ...  -48  -48  127]). Not sure if this is a quantization or scaling issue? The inputs are just 0 or 1, so the representative data set is just random 0 and 1's.

**Comment by mohantym** [comment]
Hi @gsirocco !
The genMat model, you shared in the template  is in  .txt format. I cannot load the model after changing the extension from .txt to .pkl . Would it be possible to share the sample model file again.

Thank you!

**Comment by gsirocco** [comment]
How it doesn’t accept pkl files. Maybe as a bin?

**Comment by gsirocco** [comment]
[genMat.zip](https://github.com/tensorflow/tensorflow/files/9838666/genMat.zip)
ok, I've zipped the genMat file. thanks!

**Comment by gsirocco** [comment]
Hi, just checking if you were able to access the genMat.pkl file successfully. Thanks!

**Comment by mohantym** [comment]
Hi @gsirocco !
Yeah ! I was able to replicate this issue in 2.8, 2.9 and nightly version with the model file link you shared.
Trying  to find the  scaling issue using [tfl.quantization_debugger ](https://www.tensorflow.org/lite/performance/quantization_debugger).
Thank you!

**Comment by mohantym** [comment]
@gsirocco !
Could you normalize your model using [metadata writer](https://www.tensorflow.org/lite/models/convert/metadata_writer_tutorial) and let us know the results.
Thank you!

**Comment by gsirocco** [comment]
I'm not sure what you mean by "normalize" the model? Thanks

**Comment by gsirocco** [comment]
Hi, are you just asking for us to add meta data like comments about what the model does and how it works, inputs and outputs? Thanks!

**Comment by mohantym** [comment]
@gsirocco !

Exactly, I was asking to add metadata like mean (127.5) , std (127.5) and label.txt (was not sure on this )with original lite model and save it again. 

But it might be a scaling issue too which can be found using quantization.debugger.

@sachinprasadhs !
Could you look at this issue.

Thank you!

**Comment by gsirocco** [comment]
Hi, just some feedback regarding parameters like mean and std. This model is different and simple without any training process. I assume you would need the mean, std, and labels to debug in a way for a more typical trained model?

**Comment by gsirocco** [comment]
The inputs and outputs of the model are strictly 0 and 1 values only. Hope this helps clarify. Thanks!
We look forward to any help why the TFLite model is not working like the TF model, if it's some scaling or quantization issue.

**Comment by gsirocco** [comment]
Hi @mohantym and @sachinprasadhs, just checking if there's any update on this issue? I'm trying to get this working for our product that utilizes an edge TPU, so before proper operation on the TPU is achieved, the TFLite model functionality would need to be correct. Is there a way for me to fix or debug this issue? Thanks, I appreciate any help and input on this issue as we are stuck right now.

**Comment by gsirocco** [comment]
Hi @miaout17 @sachinprasadhs @mohantym, I'm just checking if you are able to replicate the issue and see a possible workaround or fix? I am ultimately trying to run on edge TPU, but if the TFLite model is not correct then the edge TPU will not be correct. Thanks!!

**Comment by gsirocco** [comment]
Hi @miaout17 @sachinprasadhs @mohantym, I'm sorry to keep asking but I'm just looking for some feedback if there is an estimate of when and if this could be fixed please? We are at the stage in our project where a functionally working model is needed for overall operation. Please let me know so we can plan one way or another. Thank you!!

**Comment by gsirocco** [comment]
Hi @miaout17 @sachinprasadhs @mohantym, just checking for an update. Please let me know of a status!!!

**Comment by gsirocco** [comment]
Hi @miaout17 @sachinprasadhs @mohantym, checking for any progress on this issue? Thanks!

**Comment by terryheo** [comment]
@gsirocco  you've changed output type to int8 so it has int8 outputs [-128 127 127 ... -48 -48 127] instead of float outputs [0. 1. 0. ... 1. 0. 1.]. Could you share what wrong with this?

**Comment by gsirocco** [comment]
Even though the TF model has float, the input and output of the model is just 0s and 1s. The output of the TF invocation is correct and printed out ([0. 1. 0. ... 1. 0. 1.]). The TFLite invocation output is printed and is incorrect and depends on the representative dataset, which is just a series of random 0's and 1's. Depending on the representative dataset random sequence of 0's and 1's the output will change. The output should just be the same sequence of 0's and 1's as the output of the TF invocation. Thanks!!

**Comment by terryheo** [comment]
I see. One more to confirm. Is this only happening with quantization? Does float model have the similar issue?

**Comment by gsirocco** [comment]
I have not tried float, since in the end we need int because we want to convert the tflite model to run on the edge tpu which requires int. Thanks!

**Comment by terryheo** [comment]
I've tested and float model works well.

FYI, this is model dumps with `tf.lite.experimental.Analyzer.analyze(model_content=tflite_quant_model)`

float model
```
Subgraph#0 main(T#0) -> [T#6]
  Op#0 RESHAPE(T#0, T#2[-1, 14400]) -> [T#4]
  Op#1 FULLY_CONNECTED(T#4, T#3, T#-1) -> [T#5]
  Op#2 FLOOR_MOD(T#5, T#1) -> [T#6]

Tensors of Subgraph#0
  T#0(serving_default_ldpc_enc_input:0) shape_signature:[-1, 14400], type:FLOAT32
  T#1(model/tf.math.floormod/FloorMod/y) shape:[], type:FLOAT32 RO 4 bytes, buffer: 2, data:[2]
  T#2(model/tf.reshape/Reshape/shape) shape:[2], type:INT32 RO 8 bytes, buffer: 3, data:[-1, 14400]
  T#3(model/tf.linalg.matmul/MatMul1) shape:[1800, 14400], type:FLOAT32 RO 103680000 bytes, buffer: 4, data:[1, 0, 0, 0, 0, ...]
  T#4(model/tf.reshape/Reshape) shape_signature:[-1, 14400], type:FLOAT32
  T#5(model/tf.linalg.matmul/MatMul2) shape_signature:[-1, 1800], type:FLOAT32
  T#6(PartitionedCall:0) shape_signature:[-1, 1800], type:FLOAT32
```

int8 model
```
Subgraph#0 main(T#0) -> [T#8]
  Op#0 RESHAPE(T#0, T#2[-1, 14400]) -> [T#4]
  Op#1 FULLY_CONNECTED(T#4, T#3, T#-1) -> [T#5]
  Op#2 DEQUANTIZE(T#5) -> [T#6]
  Op#3 FLOOR_MOD(T#6, T#1) -> [T#7]
  Op#4 QUANTIZE(T#7) -> [T#8]

Tensors of Subgraph#0
  T#0(serving_default_ldpc_enc_input:0) shape_signature:[-1, 14400], type:INT8
  T#1(model/tf.math.floormod/FloorMod/y) shape:[], type:FLOAT32 RO 4 bytes, buffer: 2, data:[2]
  T#2(model/tf.reshape/Reshape/shape) shape:[2], type:INT32 RO 8 bytes, buffer: 3, data:[-1, 14400]
  T#3(model/tf.linalg.matmul/MatMul1) shape:[1800, 14400], type:INT8 RO 25920000 bytes, buffer: 4, data:[., ., ., ., ., ...]
  T#4(model/tf.reshape/Reshape) shape_signature:[-1, 14400], type:INT8
  T#5(model/tf.linalg.matmul/MatMul2) shape_signature:[-1, 1800], type:INT8
  T#6(tfl.dequantize) shape_signature:[-1, 1800], type:FLOAT32
  T#7(PartitionedCall:01) shape_signature:[-1, 1800], type:FLOAT32
  T#8(PartitionedCall:0) shape_signature:[-1, 1800], type:INT8

```


I think it's related with quantization spec from representative_dataset. Since the input range is known, it might be better to provide it manually.

@yyoon any idea on this?

**Comment by gsirocco** [comment]
Thanks, I just confirmed that float works as well. How would this be provided manually? The range is just 0 to 1 and only 0 and 1 not in between. Thanks!!

**Comment by gsirocco** [comment]
Hi, checking if there's anything that can be tried? I'm not sure how to provide the input range manually. Thanks!

**Comment by gsirocco** [comment]
Hi, I was wondering if there's another avenue that I should pursue to get assistance with this issue? It is a TFLite converter issue correct? Thanks

**Comment by gsirocco** [comment]
Hi, I'm checking if there's a way to fix this issue or to go through some other source? Or some other information needs to be provided? Thanks

**Comment by gsirocco** [comment]
Hi, is the best approach to debug this to go through the tflite converter process or to go through the invoke and see what the issue is?

**Comment by gsirocco** [comment]
Hi @yyoon @miaout17 , just checking if there's any update on what can be done for this issue. Thanks!!

**Comment by gsirocco** [comment]
Hi, we really want to productize this utilizing a TPU but if we can't find a way to do this then we'll have to try to find alternatives. Please let us know what can be done to get this working. Thanks and happy holidays!!

**Comment by gsirocco** [comment]
Hi, we are really stuck with this issue. We can only utilize the Edge TPU for this application, regular Tensorflow or TFLite won't provide the performance that is needed for this application, so unless we can convert to using int's in TFLite models properly then the TPU can't be utilized. Thank you to anyone who can help

**Comment by gsirocco** [comment]
Hi, is there any channel for help besides this that I can reach out to fix this TFLite conversion issue?
This is critical for getting our product to work for performance using TPU
Thank you!!!

**Comment by gsirocco** [comment]
Hi, checking if there is any path forward that I can take to fix this? Thanks!

**Comment by haozha111** [comment]
@LukeBoyer could you help with this issue? thanks!

**Comment by LukeBoyer** [comment]
@haozha111 Sure, I'll take a look asap

**Comment by gsirocco** [comment]
Thank you!! Please let me know if you have any questions or need anymore info!!

**Comment by gsirocco** [comment]
Hi Just checking if you were able to duplicate the issue? Thanks!

**Comment by gsirocco** [comment]
Hi @LukeBoyer, just checking if you are able to duplicate the issue and have any idea what the problem may be. Thanks!

**Comment by gsirocco** [comment]
Great, thanks @LukeBoyer !!

**Comment by LukeBoyer** [comment]
Hi @gsirocco, I can confirm I can reproduce this by running the supplied code. Just to confirm, are you receiving this same warning?
`tensorflow/lite/python/convert.py:789: UserWarning: Statistics for quantized inputs were expected, but not specified; continuing anyway.`

**Comment by gsirocco** [comment]
Hi @LukeBoyer, yes I am seeing the same message. I assume this is part of the problem? How do I correct this?
I am specifying a representative dataset which is just random 1's and 0's (binary only). Is there something else that needs to be specified somehow? Thanks!

**Comment by gsirocco** [comment]
Hi @LukeBoyer, did you have any suggestions regarding the error, maybe specifying statistics for the quantized inputs somehow? In this case it would simply be 50% 0 and 50% 1. Thanks!

**Comment by LukeBoyer** [comment]
Hey @gsirocco I have some time to investigate this week and send an update on Friday.

**Comment by LukeBoyer** [comment]
Hi. Sorry for the delay. I see your point in that the it would make sense for the output to be 0s and 1s as you said. If we take a closer look at the model it looks like the last operation is a `tfl.quantize` which is outputting the observed data across the int8 range.

could you try omitting this line and see if it resolves your issue? `converter.inference_output_type = tf.int8  # or tf.uint8`

Also, this documentation is an invaluable for understanding the quantization workflow https://www.tensorflow.org/lite/performance/post_training_quantization.

**Comment by gsirocco** [comment]
Hi @LukeBoyer, thanks for the suggestion. I have tried this, but it did not fix the issue. What I am noticing is the output depends on the representative_dataset that is provided. In this case, I have just been generating a series of random 1's and 0's which will differ with each run. If I do this with TFLite conversion that is float32 then correct output is generated, however with int8 conversion the output is incorrect. I will take a look at the link that you have provided and look forward to any other suggestions you may have. 

I have placed updated code that compares the 1800 output len to the last 1800 entries of the output file. Thanks!

```
import tensorflow as tf
from tensorflow import keras
import pickle
import numpy as np

print(tf.__version__)

with open("genMat.pkl", "rb") as f:
     gen_matrix = pickle.load(f)
# 14400x1800
xlen=14400
ylen=1800

def representative_dataset():
  for _ in range(100):
      data = np.random.randint(0, 2, (xlen,))
      yield [data.astype(np.float32)]

test_input = np.loadtxt("input_msg.csv",dtype=int,delimiter=',')
test_output = np.loadtxt("output_msg.csv",dtype=int,delimiter=',')
input = 0
generator = tf.convert_to_tensor(gen_matrix[input*xlen:(input+1)*xlen, :], dtype=tf.float32)

ldpc_enc_input = keras.Input(shape=(xlen,), name="ldpc_enc_input")
generator_input = tf.reshape(ldpc_enc_input, shape=(-1, xlen))
generator_output = tf.matmul(generator_input, generator)

ldpc_enc_output = tf.math.floormod(generator_output, 2)

ldpc_enc_model = keras.Model(inputs=ldpc_enc_input, outputs=ldpc_enc_output)

arr=ldpc_enc_model(test_input[input*xlen:(input+1)*xlen])
tfout = arr.numpy()[0]
print(tfout)
if(test_output[-1800:] == tfout).all():
     print("TF output is correct!!!")
else:
     print("TF output is BAD!!!!!")


converter = tf.lite.TFLiteConverter.from_keras_model(ldpc_enc_model)

converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.representative_dataset = representative_dataset
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.target_spec.supported_types = [tf.int8]
converter.inference_input_type = tf.int8 # or tf.uint8
#converter.inference_output_type = tf.int8 # or tf.uint8

tflite_quant_model = converter.convert()
'''
fileName = 'regmodel.tflite'
with open(fileName, 'wb') as f:
     f.write(tflite_quant_model)
'''
# Load the TFLite model and allocate tensors.
interpreter = tf.lite.Interpreter(model_content=tflite_quant_model)
interpreter.allocate_tensors()

# Get input and output tensors.
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

input_shape = input_details[0]['shape']
input_data = np.array(test_input[input*xlen:(input+1)*xlen], dtype=np.int8, ndmin=2)
interpreter.set_tensor(input_details[0]['index'], input_data)
#test_input = np.loadtxt("input_msg.csv",dtype=np.float32,delimiter=',')
#interpreter.set_tensor(input_details[0]['index'], np.array(test_input[input*xlen:(input+1)*xlen], dtype=np.float32, ndmin=2))

interpreter.invoke()

# The function `get_tensor()` returns a copy of the tensor data.
# Use `tensor()` in order to get a pointer to the tensor.
arr = interpreter.get_tensor(output_details[0]['index'])
tflout = arr[0]
print(tflout)
#np.round(tflout)
for i in range(len(tflout)):
     if tflout[i] < 0.5:
          tflout[i] = 0
     else:
          tflout[i] = 1
print(tflout)

if(test_output[-1800:] == tflout).all():
     print("TFLite output is correct!!!")
else:
     print("TFLite output is BAD!!!!!")
print(arr[0])
```

**Comment by gsirocco** [comment]
Btw, I have attached the output file with the correct output, which I don't think I provided previously.
[output_msg.csv](https://github.com/tensorflow/tensorflow/files/10541733/output_msg.csv)

**Comment by yyoon** [comment]
Thanks @terryheo for routing this to me, and sorry @gsirocco for my late reply. Somehow this issue fell off my radar.

After reading through what has been discussed so far, I don't think there's anything wrong happening. I believe @terryheo's initial observation is right, and AFAICT, in your very first code in https://github.com/tensorflow/tensorflow/issues/58206#issue-1417553857, the model is correctly being quantized using the full-integer int8 quantization scheme. Let me try my best to explain.

What you need to understand that when you do full-integer quantization, the output values you get are the quantized values, and not the real values. With int8 quantization, the min value would map to `-128`, and max to `127`. Since the expected float version outputs are just `0`s and `1`s, roughly, you could imagine that `0` (real value) would be represented as `-128` (quantized value) and `1` (real value) as `127` (quantized value).

The only thing you need to do after getting the quantized values is to dequantize them.

Given this formula: `real_value = (int8_value - zero_point) * scale` (see: https://www.tensorflow.org/lite/performance/quantization_spec#specification_summary),
you could add the following code to print the dequantized float values.

```
output_scale, output_zero_point = output_details[0]['quantization']
print((arr[0].astype(float) - output_zero_point) * output_scale)
```

An example run from my side prints like this:

```
[-128  127  127 ...  -81  127  127]
[0.         1.00000006 1.00000006 ... 0.18431374 1.00000006 1.00000006]
```

You can see that `-128` from the output tensor maps to real value `0`, and `127` from the output maps to `1` (with some floating point error margin). Hope this helps.

**Comment by gsirocco** [comment]
Hi @yyoon thanks for the feedback. It makes sense that the output would need to be scaled back to 0 and 1 values.
However, when I do this scaling to get 0 and 1 values, they are not correct with respect to the expected values and Tensorflow output. I have provided a couple of example output runs showing the first 10 entries of output. It seems that the output values depend on the representative dataset that is provided, which is random 0s and 1s. I have included the sample code to generate and compare below. 
Please let me know if this makes sense or not. Thanks!

```
TF output is correct!!!
correct output:  [0 1 0 0 0 1 1 0 1 0]
TF output:  [0. 1. 0. 0. 0. 1. 1. 0. 1. 0.]
TFLite output:  [-128  127  -38  -38  127  127   52   52  127  127]
TFLitescale output:  [0.         1.00000006 0.3529412  0.3529412  1.00000006 1.00000006
 0.70588239 0.70588239 1.00000006 1.00000006]
TFLiteint output:  [0, 1, 0, 0, 1, 1, 1, 1, 1, 1]
TFLite output is BAD!!!!!
```

```
TF output is correct!!!
correct output:  [0 1 0 0 0 1 1 0 1 0]
TF output:  [0. 1. 0. 0. 0. 1. 1. 0. 1. 0.]
TFLite output:  [-128  127  127  -26  -26  127  127   76   76 -128]
TFLitescale output:  [0.         1.00000006 1.00000006 0.40000002 0.40000002 1.00000006
 1.00000006 0.80000005 0.80000005 0.        ]
TFLiteint output:  [0, 1, 1, 0, 0, 1, 1, 1, 1, 0]
TFLite output is BAD!!!!!
```

```
import tensorflow as tf
from tensorflow import keras
import pickle
import numpy as np

print(tf.__version__)

with open("genMat.pkl", "rb") as f:
     gen_matrix = pickle.load(f)
# 14400x1800
xlen=14400
ylen=1800

def representative_dataset():
  for _ in range(100):
      data = np.random.randint(0, 2, (xlen,))
      yield [data.astype(np.float32)]

test_input = np.loadtxt("input_msg.csv",dtype=int,delimiter=',')
test_output = np.loadtxt("output_msg.csv",dtype=int,delimiter=',')
tfoutcorr=test_output[-1800:]
input = 0
generator = tf.convert_to_tensor(gen_matrix[input*xlen:(input+1)*xlen, :], dtype=tf.float32)

ldpc_enc_input = keras.Input(shape=(xlen,), name="ldpc_enc_input")
generator_input = tf.reshape(ldpc_enc_input, shape=(-1, xlen))
generator_output = tf.matmul(generator_input, generator)

ldpc_enc_output = tf.math.floormod(generator_output, 2)

ldpc_enc_model = keras.Model(inputs=ldpc_enc_input, outputs=ldpc_enc_output)

arr=ldpc_enc_model(test_input[input*xlen:(input+1)*xlen])
tfout = arr.numpy()[0]

if(test_output[-1800:] == tfout).all():
     print("TF output is correct!!!")
else:
     print("TF output is BAD!!!!!")


converter = tf.lite.TFLiteConverter.from_keras_model(ldpc_enc_model)

converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.representative_dataset = representative_dataset
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.target_spec.supported_types = [tf.int8]
converter.inference_input_type = tf.int8 # or tf.uint8
converter.inference_output_type = tf.int8 # or tf.uint8

tflite_quant_model = converter.convert()

# Load the TFLite model and allocate tensors.
interpreter = tf.lite.Interpreter(model_content=tflite_quant_model)
interpreter.allocate_tensors()

# Get input and output tensors.
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

input_shape = input_details[0]['shape']
input_data = np.array(test_input[input*xlen:(input+1)*xlen], dtype=np.int8, ndmin=2)
interpreter.set_tensor(input_details[0]['index'], input_data)

interpreter.invoke()

# The function `get_tensor()` returns a copy of the tensor data.
# Use `tensor()` in order to get a pointer to the tensor.
arr = interpreter.get_tensor(output_details[0]['index'])
tflout = arr[0]
output_scale, output_zero_point = output_details[0]['quantization']
tfloutscale = (arr[0].astype(float) - output_zero_point) * output_scale

tfloutint=[0]*1800
for i in range(len(tflout)):
     if tfloutscale[i] < 0.5:
          tfloutint[i] = 0
     else:
          tfloutint[i] = 1

print("correct output: ",tfoutcorr[0:10])
print("TF output: ",tfout[0:10])
print("TFLite output: ",tflout[0:10])
print("TFLitescale output: ",tfloutscale[0:10])
print("TFLiteint output: ",tfloutint[0:10])

if(test_output[-1800:] == tfloutint).all():
     print("TFLite output is correct!!!")
else:
     print("TFLite output is BAD!!!!!")
```

**Comment by gsirocco** [comment]
Hi @yyoon and @LukeBoyer,  I have been looking at the [quantization debugger](https://www.tensorflow.org/lite/performance/quantization_debugger) and am wondering if it would be applicable in this case to understand if and where a quantization error is coming from in the converted model?
Thanks!

**Comment by gsirocco** [comment]
I have run the quantization debugger on the converted integer TFLite model 
```
debugger = tf.lite.experimental.QuantizationDebugger(
    converter=converter, debug_dataset=rep_dat())
debugger.run()
RESULTS_FILE = '/home/gsosnow/doc/debugger_results.csv'
with open(RESULTS_FILE, 'w') as f:
  debugger.layer_statistics_dump(f)
```
and have gotten the output in the attached csv file. I'm not sure if this is complete and done correctly.
I'm not sure how to interpret this output per the debugger link but the second entry (FULLY_CONNECTED MatMul2) has relatively large values, not sure if it's significant.
[debugger_results.csv](https://github.com/tensorflow/tensorflow/files/10613818/debugger_results.csv)

Trying to analyze the data for range and RMSE/scale:
op_name                   mean_squared_error            scale           range=(255*scale)    sqrt(mean_squared_error)/scale   tensor_name
RESHAPE                          0                               0.003921569              1.00011                         0                                     Reshape
FULLY_CONNECTED       53.292534                25.309803               6453.999                            0.28843265                     MatMul2

**Comment by gsirocco** [comment]
Hi @yyoon and @LukeBoyer, just checking if you guys verify that the results of the integer quantization, whether in their quantized state or dequantized are incorrect vs. the floating point output of the TFLite model and TF output? 
I'm not sure if using the quantization debugger would reveal the source of error? Thanks!

**Comment by gsirocco** [comment]
Hi @LukeBoyer, earlier you had pointed out the warning that was getting printed out in conversion of the TF model to TFLite int8 `tensorflow/lite/python/convert.py:789: UserWarning: Statistics for quantized inputs were expected, but not specified; continuing anyway.` 

I set a breakpoint just before this location, and manually supplied the statistics, mean and std both to 0.5 because the input values are only 0's and 1's. I did this by setting `quantized_input_stats=[(0.5,0.5)]` since there was no apparent way to supply this info from the `converter.convert()` function call. 
However, this did not correct the output. It was still be the same output.

I will take a deeper look at the quantization debugger and it's output to see if it will provide any clues to the issue. Thanks!

**Comment by gsirocco** [comment]
One thing I have tried is to change from int8 to uint8 since the input is just 0's and 1's this should still work.
I do this by changing the `inference_input_type` and `inference_output_type` to `tf.uint8` as below.
However the output in this case is the first 319 entries are 0's and the final 1481 entries are 255. 
I'm not sure why this would be the case.

```
converter = tf.lite.TFLiteConverter.from_keras_model(ldpc_enc_model)

converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.representative_dataset = representative_dataset
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.target_spec.supported_types = [tf.int8]
converter.inference_input_type = tf.uint8
converter.inference_output_type = tf.uint8
tflite_quant_model = converter.convert()
```

```
correct output:  [0 1 0 0 0 1 1 0 1 0]
TF output:  [0. 1. 0. 0. 0. 1. 1. 0. 1. 0.]
TFLite output:  [0 0 0 0 0 0 0 0 0 0]
TFLitescale output:  [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
TFLiteint output:  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
TFLite output is BAD!!!!!
```

**Comment by gsirocco** [comment]
Hi @yyoon and @LukeBoyer, I am trying to run the quantization debugger to narrow down the source of quantization error by eliminating specific nodes or ops from quantization as described [in](https://www.tensorflow.org/lite/performance/quantization_debugger) by doing selective quantization.
However, when I try to do this, I get the error `Please check if the quantized model is in debug mode`
From the documentation, I don't see how to get the model into 'debug mode' or what the criteria are for the model being in debug mode. Below is the code I am trying to run for the model to do the quantization debugging and selective quantization.
It fails when the debug options are provided in the second call to instantiating the QunatizationDebugger.
Any pointers of how to get past this are appreciated! Thanks!!

```
suspected_layers = ['model/tf.reshape/Reshape', 'model/tf.linalg.matmul/MatMul2']
suspected_ops = ['RESHAPE', 'FULLY_CONNECTED']
debug_options = tf.lite.experimental.QuantizationDebugOptions(
     denylisted_ops=suspected_ops)
#    denylisted_nodes=suspected_layers)
debugger = tf.lite.experimental.QuantizationDebugger(
    converter=converter,
    debug_dataset=rep_dat())
selective_quantized_model = debugger.get_debug_quantized_model()
debugger = tf.lite.experimental.QuantizationDebugger(
    converter=converter,
    debug_dataset=rep_dat(),
    debug_options=debug_options)
selective_quantized_model = debugger.get_nondebug_quantized_model()
```

**Comment by gsirocco** [comment]
Hi @yyoon and @LukeBoyer, I am stuck on the quantization debugger because of the debug mode issue. So I found the process called quantization aware training described [here](https://www.tensorflow.org/model_optimization/guide/quantization/training_example). I tried to run this as below to create a quantized model based on test data, however I am getting the following error which I'm not sure how to fix exactly. Thanks!!
```
RuntimeError: Layer tf.reshape:<class 'keras.layers.core.tf_op_layer.TFOpLambda'> is not supported. You can quantize this layer by passing a `tfmot.quantization.keras.QuantizeConfig` instance to the `quantize_annotate_layer` API.
```

```
import tensorflow_model_optimization as tfmot

quantize_model = tfmot.quantization.keras.quantize_model
# q_aware stands for for quantization aware.
q_aware_model = quantize_model(ldpc_enc_model)
```

**Comment by gsirocco** [comment]
Hi @miaout17 @sachinprasadhs @yyoon @LukeBoyer @haozha111, is it better to raise this quantization debugger issue in a new issue? I think this is the best path to narrow down the source of quantization error. Although once narrowed down, in order to fix the issue, the quantization aware training may be best to fix it? Please let me know what would be the best path. I am desperate in trying to solve this. Thanks!

> Hi @yyoon and @LukeBoyer, I am trying to run the quantization debugger to narrow down the source of quantization error by eliminating specific nodes or ops from quantization as described [in](https://www.tensorflow.org/lite/performance/quantization_debugger) by doing selective quantization. However, when I try to do this, I get the error `Please check if the quantized model is in debug mode` From the documentation, I don't see how to get the model into 'debug mode' or what the criteria are for the model being in debug mode. Below is the code I am trying to run for the model to do the quantization debugging and selective quantization. It fails when the debug options are provided in the second call to instantiating the QunatizationDebugger. Any pointers of how to get past this are appreciated! Thanks!!
> 
> ```
> suspected_layers = ['model/tf.reshape/Reshape', 'model/tf.linalg.matmul/MatMul2']
> suspected_ops = ['RESHAPE', 'FULLY_CONNECTED']
> debug_options = tf.lite.experimental.QuantizationDebugOptions(
>      denylisted_ops=suspected_ops)
> #    denylisted_nodes=suspected_layers)
> debugger = tf.lite.experimental.QuantizationDebugger(
>     converter=converter,
>     debug_dataset=rep_dat())
> selective_quantized_model = debugger.get_debug_quantized_model()
> debugger = tf.lite.experimental.QuantizationDebugger(
>     converter=converter,
>     debug_dataset=rep_dat(),
>     debug_options=debug_options)
> selective_quantized_model = debugger.get_nondebug_quantized_model()
> ```

**Comment by gsirocco** [comment]
Hi, I'm just checking if there's any feedback on any of these issues, even just the incorrectness of the dequantized output of the int8 model or the issues of trying to run the quantization debugger to isolate which node a quantization error is coming from. This is really preventing us from being able to get the performance for practical implementation and hardware like the TPU. I don't see any alternatives and am really stuck here. Thanks for any response and help!

**Comment by gsirocco** [comment]
Hi, checking on any thoughts or suggestions. I look forward to any feedback or help. Thanks!

**Comment by gsirocco** [comment]
Hi, is there another place that I should look for assistance with this issue? Maybe there's a more pertinent place to ask for help on this? Thanks!

**Comment by gsirocco** [comment]
Hi, looking for any help or suggestions. This is very close to working for our product if only can get the correct answer. The TPU is providing the speed that just doing TF or TFLite can't. It's discouraging that I can't get over this last hurdle.

**Comment by gsirocco** [comment]
Hi @miaout17, @yyoon, @haozha111, @LukeBoyer, and @sachinprasadhs,
I've been struggling with this issue for almost 5 months now and I fear it is a dead end?
I have tried to utilize the tools such as quantization debugger and quantization aware training to get to the source of this issue, but these tools themselves seem to have some issue. If you have any suggestion on these tools or any other for me to debug and fix this then I would really appreciate it. As I mentioned before, for performance and cost reasons, the TPU is the only viable way to implement this product, but trying to overcome this issue is preventing this. Please help!! Thank you!!

**Comment by gsirocco** [comment]
Hi, looking for any suggestions on the quantization debugger and/or quantization aware training to figure out this issue, or any other way that you know of. It's been frustrating to try to resolve this. I wish there were better response. Is there something that I need to do differently to get help then please let me know. This system is very frustrating and I feel like I have failed

**Comment by yyoon** [comment]
Apologies for the delayed response. Before we jump in to the details of quantization debugger, wanted to do a quick sanity check. Are you sure you're correctly quantizing your inputs before feeding them to the model?

In https://github.com/tensorflow/tensorflow/issues/58206#issuecomment-1416118721 I don't see any code for getting the quantization parameters for the input tensor, which would be causing issues.

**Comment by gsirocco** [comment]
Hi @yyoon, thanks for the response. In terms of quantizing the input, it's already just 0's and 1's, which are put into the input tensor as below. Is there some scaling of this input data that should be done before setting it as the input tensor?

```
input_data = np.array(test_input[input*xlen:(input+1)*xlen], dtype=np.int8, ndmin=2)
interpreter.set_tensor(input_details[0]['index'], input_data)
```
Here is a subset of the raw input data and then what is set for the input tensor:
```
test_input[0:32]
array([1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0,
       0, 1, 0, 0, 0, 1, 1, 0, 1, 0])
input_data[0][0:32]
array([1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0,
       0, 1, 0, 0, 0, 1, 1, 0, 1, 0], dtype=int8)
```

**Comment by yyoon** [comment]
I think the representative dataset should stay in its original format (0, 1 values), whereas the input_data needs to be quantized based on the input scale and input zero point value.

See https://www.tensorflow.org/lite/performance/post_training_integer_quant#run_the_tensorflow_lite_models for an example.

**Comment by gsirocco** [comment]
Thanks for the clarification and example. I think I have done this correctly. However, it still looks like the output isn't right.

```
def representative_dataset():
  for _ in range(100):
      data = np.random.randint(0, 2, (xlen,))
      yield [data.astype(np.float32)]

input_shape = input_details[0]['shape']
input_scale, input_zero_point = input_details[0]["quantization"]
for i in range(len(test_input)):
     test_input[i] = test_input[i] / input_scale + input_zero_point
input_data = np.array(test_input[input*xlen:(input+1)*xlen], dtype=np.int8, ndmin=2)
interpreter.set_tensor(input_details[0]['index'], input_data)

interpreter.invoke()

arr = interpreter.get_tensor(output_details[0]['index'])
tflout = arr[0]
output_scale, output_zero_point = output_details[0]['quantization']
tfloutscale = (arr[0].astype(float) - output_zero_point) * output_scale
print("correct output: ",tfoutcorr[0:10])
print("TF output: ",tfout[0:10])
print("TFLite output: ",tflout[0:10])
print("TFLitescale output: ",tfloutscale[0:10])
```
```
correct output:  [0 1 0 0 0 1 1 0 1 0]
TF output:  [0. 1. 0. 0. 0. 1. 1. 0. 1. 0.]
TFLite output:  [-128  127  127    4    4  127  127  127  127  -53]
TFLitescale output:  [0.         1.00000006 1.00000006 0.51764709 0.51764709 1.00000006
 1.00000006 1.00000006 1.00000006 0.29411766]
TFLiteint output:  [0, 1, 1, 1, 1, 1, 1, 1, 1, 0]
```

**Comment by gsirocco** [comment]
Would this indicate that the quantization debugger is the route to go in order to debug this? Thanks!!

**Comment by gsirocco** [comment]
Hi, just checking if there's any suggestion for next step on this issue? Thanks!!

**Comment by gsirocco** [comment]
Hi, checking in for any help this. Thanks!

**Comment by gsirocco** [comment]
I feel like using the quantization debugger would be the next right step??

**Comment by gsirocco** [comment]
Hi anyone have any suggestions on this? Thanks!
It's been a real struggle, please help!!

**Comment by pkgoogle** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The TFLite team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the TFLite space.
